---
title: "GenePackage_Vignette"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{GenePackage_Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, include=FALSE}
library(GenePackage)

if (!requireNamespace("GenomicRanges", quietly = TRUE)) {
  install.packages("BiocManager")
  BiocManager::install("GenomicRanges")
}
library(GenomicRanges)
```

## Introduction

The GenePackage is designed to provide a structured and consistent way to represent different types of genes including protein-coding genes, long non-coding RNA (lncRNA) genes, microRNA (miRNA) genes, small nuclear RNA (snRNA) genes, ribosomal RNA (rRNA) genes, transfer RNA (tRNA) genes, pseudogenes and enhancer RNA (eRNA) genes. It's based on S4 classes in R to ensure that gene-related data is well-organized, making it easier to handle genomic data.

## Class Structure

### Gene Class

The Gene class is a virtual base class that provides a generic structure for representing a gene.It includes the following slots:
* id: Gene ID (character)
* hugoSymbol: HUGO Symbol (character)
* name : Gene name (character)
* description : Description of the gene (character)
* structure : Gene structure, as a GRanges object


### ProteinCodingGene Class

This class extends the Gene class and is designed to represent protein-coding genes. It includes additionally slots:
* proteinID : A unique identifier for the protein product encoded by the gene (character)
* proteinSequence : The amino acid sequence of the protein encoded by the gene (character)
* exonCount : The number of exons in the gene (integer)

``` {r example_ProteinCodingGene}
gr <- GRanges(seqnames = "chr1", ranges = IRanges(start = 1000, end = 5000))
protein_gene <- new("ProteinCodingGene",
                    id = "ENSG00000139618",
                    hugoSymbol = "BRCA2",
                    name = "BRCA2",
                    description = "Breast cancer type 2 susceptibility protein",
                    structure = gr,
                    proteinID = "NP_000050.2",
                    proteinSequence = "MSSQLQALFQS...",
                    exonCount = as.integer(27))
protein_gene
```

### LongNonCodingRNAGene Class

This class represents long non-coding RNA (lncRNA) genes and extends the Gene class. It includes:
* lncRNAID : A unique identifier for the lncRNA product (character)
* rnaSequence : The nucleotide sequence of the RNA encoded by the lncRNA gene (character)

``` {r example_lncRNA}
gr <- GRanges(seqnames = "chr11", ranges = IRanges(start = 100000, end = 110000))
lncrna_gene <- new("LongNonCodingRNAGene",
                   id = "ENSG00000224207",
                   hugoSymbol = "HOTAIR",
                   name = "HOTAIR",
                   description = "HOX Transcript Antisense RNA",
                   structure = gr,
                   lncRNAID = "lncHOTAIR",
                   rnaSequence = "AUGCUCGAA...")
lncrna_gene
```

### MicroRNAGene Class

This class represents microRNA (miRNA) genes and extends the Gene class. It includes:

* miRNAID : A unique identifier for the miRNA product (character)
* rnaSequence : The nucleotide sequence of the RNA encoded by the miRNA gene (character)
* seedSequence : The seed sequence of the miRNA, which is crucial for its function (character)

``` {r example_miRNA}
gr <- GRanges(seqnames = "chr7", ranges = IRanges(start = 30000, end = 31000))
mirna_gene <- new("MicroRNAGene",
                  id = "ENSG00000212351",
                  hugoSymbol = "MIR21",
                  name = "miR-21",
                  description = "MicroRNA 21, involved in cancer progression",
                  structure = gr,
                  miRNAID = "miR-21",
                  rnaSequence = "AAGCUUUC...",
                  seedSequence = "AAGCUU")
mirna_gene
```

### SmallNuclearRNAGene Class

This class represents small nuclear RNA (snRNA) genes and extends the Gene class. It includes: 

* snRNAID : A unique identifier for the small nuclear RNA product (character).
* rnaSequence : The nucleotide sequence of the RNA encoded by the snRNA gene (character).

``` {r example_snRNA}
gr <- GRanges(seqnames = "chr1", ranges = IRanges(start = 2000, end = 3000))
snrna_gene <- SmallNuclearRNAGene(
  id = "ENSG00000212345",
  hugoSymbol = "U1",
  name = "U1 small nuclear RNA",
  snRNAID = "SNRNP70",
  rnaSequence = "AUGCUAGC...",
  description = "U1 snRNA is involved in the splicing of pre-mRNA.",
  structure = gr
)
snrna_gene
```

### RibosomalRNAGene Class

This class represents ribosomal RNA (rRNA) genes and extends the Gene class. It includes: 

* rRNAID : A unique identifier for the ribosomal RNA product (character).
* rnaSequence : The nucleotide sequence of the RNA encoded by the rRNA gene (character).

``` {r example_rRNA}
gr <- GRanges(seqnames = "chr13", ranges = IRanges(start = 100000, end = 101000))
rrna_gene <- RibosomalRNAGene(
  id = "ENSG00000212346",
  hugoSymbol = "18S",
  name = "18S ribosomal RNA",
  rRNAID = "RRNA18S",
  rnaSequence = "AGCUAGCU...",
  description = "18S rRNA is a component of the small ribosomal subunit.",
  structure = gr
)
rrna_gene
```

### TransferRNAGene Class

This class represents transfer RNA (tRNA) genes and extends the Gene class. It includes: 

* tRNAID : A unique identifier for the transfer RNA product (character).
* aminoAcid : The amino acid carried by the tRNA (character).
* rnaSequence : The nucleotide sequence of the RNA encoded by the tRNA gene (character).

``` {r example_tRNA}
gr <- GRanges(seqnames = "chr6", ranges = IRanges(start = 50000, end = 50500))
trna_gene <- TransferRNAGene(
  id = "ENSG00000212347",
  hugoSymbol = "TRNA",
  name = "Transfer RNA for Alanine",
  tRNAID = "TRNAAla",
  aminoAcid = "Alanine",
  rnaSequence = "GGGUUACCC...",
  description = "This tRNA carries alanine to the ribosome during translation.",
  structure = gr
)
trna_gene
```

### Pseudogene Class

This class represents pseudogenes and extends the Gene class. It includes: 

* parentGeneID : The unique identifier for the parent gene from which the pseudogene originated (character).
* functionalStatus : The functional status of the pseudogene, indicating whether it is transcribed, translated, or non-functional (character).

``` {r example_pseudogenes}
gr <- GRanges(seqnames = "chr11", ranges = IRanges(start = 300000, end = 305000))
pseudo_gene <- Pseudogene(
  id = "ENSG00000212348",
  hugoSymbol = "Î¨BRCA1",
  name = "Pseudogene of BRCA1",
  parentGeneID = "ENSG00000012048",
  functionalStatus = "non-functional",
  description = "This is a pseudogene derived from the BRCA1 gene.",
  structure = gr
)
pseudo_gene
```


### EnhancerRNAGene Class

The EnhancerRNAGene class represents enhancer RNA (eRNA) genes and extends the Gene class. It includes: 

* targetGeneID : The unique identifier for the gene targeted by the enhancer RNA (character).
* rnaSequence : The nucleotide sequence of the RNA encoded by the eRNA gene (character).

``` {r example_eRNA}
gr <- GRanges(seqnames = "chr8", ranges = IRanges(start = 150000, end = 151000))
erna_gene <- EnhancerRNAGene(
  id = "ENSG00000212349",
  hugoSymbol = "eRNA",
  name = "Enhancer RNA near MYC",
  targetGeneID = "ENSG00000136997",
  rnaSequence = "AUCGUAUC...",
  description = "This eRNA is associated with the regulation of the MYC gene.",
  structure = gr
)
erna_gene
```

##Functions and methods 
These functions allow you to easily retrieve or modify the HUGO Symbol associated with a gene.
### Getting the HUGO Symbol
The `hugoSymbol` function retrieves the HUGO Symbol of a gene object. This function works with any object that is a subclass of Gene.
``` {r example_getter}
hugoSymbol(protein_gene)
```

### Setting the HUGO Symbol
The `hugoSymbol<-` function allows you to modify the HUGO Symbol for a gene object.
``` {r example_setter}
hugoSymbol(protein_gene) <- "NEW_SYMBOL"
hugoSymbol(protein_gene)
```

### Calculating the Length of Gene Products
The lengthProduct function is a generic function that calculates the length of the gene product, a protein or RNA sequence, depending on the type of gene.

``` {r example_lengthProduct}
lncrna_gene <- LongNonCodingRNAGene(
  id = "ENSG00000212350",
  hugoSymbol = "XIST",
  name = "XIST",
  lncRNAID = "lncXIST",
  rnaSequence = "AUGCUAGC...",
  description = "X-inactive specific transcript",
  structure = GRanges(seqnames = "chrX", ranges = IRanges(start = 5000, end = 20000))
)
lengthProduct(lncrna_gene)
```

## Conclusion
The GenePackage provides a robust framework for representing various gene types and their associated data in a structured manner using S4 classes. This package is particularly useful for bioinformatics tasks that require a detailed representation of genomic data.

```{r session-info, echo=FALSE}
sessionInfo()
```
